scripts/umap_reducer.py

```diff
@@ -39,71 +39,100 @@ def main(fmat, fxyz, ftags, fcolor, colorscol, prefix, output, peratom, keepraw,
-    use_atomic_desc = (peratom or plotatomic or projectatomic)
+    if not use_morgan:
+
+        foutput = prefix + "-umap-d" + str(umap_d)
+        use_atomic_desc = (peratom or plotatomic or projectatomic)
+
+        # try to read the xyz file
+        if fxyz != 'none':
+            asapxyz = ASAPXYZ(fxyz)
+            desc, desc_atomic = asapxyz.get_descriptors(fmat, use_atomic_desc)
+            if projectatomic:
+                desc = desc_atomic.copy()
+        else:
+            asapxyz = None
+            print("Did not provide the xyz file. We can only output descriptor matrix.")
+            output = 'matrix'
+        # we can also load the descriptor matrix from a standalone file
+        if os.path.isfile(fmat):
+            try:
+                desc = np.genfromtxt(fmat, dtype=float)
+                print("loaded the descriptor matrix from file: ", fmat)
+            except:
+                raise ValueError('Cannot load the descriptor matrix from file')
+        # sanity check
+        if len(desc) == 0:
+            raise ValueError('Please supply descriptor in a xyz file or a standlone descriptor matrix')
+        print("shape of the descriptor matrix: ", np.shape(desc), "number of descriptors: ", np.shape(desc[0]))
+
+        if ftags != 'none':
+            tags = np.loadtxt(ftags, dtype="str")[:]
+            ndict = len(tags)
+
+        # scale & center
+        if scale:
+            from sklearn.preprocessing import StandardScaler
+            scaler = StandardScaler()
+            print('Shape of descriptor matrix is {}'.format(desc.shape))
+            print(scaler.fit(desc))
+            desc = scaler.transform(desc)  # normalizing the features
+
+        # fit UMAP
+
+        reducer = umap.UMAP(n_neighbors=5)
+        proj = reducer.fit_transform(desc)
+        if peratom or plotatomic and not projectatomic:
+            proj_atomic_all = reducer.transform(desc_atomic)
+
+        # save
+        if output == 'matrix':
+            np.savetxt(foutput + ".coord", proj, fmt='%4.8f', header='low D coordinates of samples')
+            if peratom:
+                np.savetxt(foutput + "-atomic.coord", proj_atomic_all, fmt='%4.8f', header='low D coordinates of samples')
+        if output == 'xyz':
+            if os.path.isfile(foutput + ".xyz"):
+                os.rename(foutput + ".xyz", "bck." + foutput + ".xyz")
+            asapxyz.set_descriptors(proj, 'umap_coord')
+            if peratom:
+                asapxyz.set_atomic_descriptors(proj_atomic_all, 'umap_coord')
+            # remove the raw descriptors
+            if not keepraw:
+                asapxyz.remove_descriptors(fmat)
+                asapxyz.remove_atomic_descriptors(fmat)
+            asapxyz.write(foutput)
 
-    # try to read the xyz file
-    if fxyz != 'none':
-        asapxyz = ASAPXYZ(fxyz)
-        desc, desc_atomic = asapxyz.get_descriptors(fmat, use_atomic_desc)
-        if projectatomic: desc = desc_atomic.copy()
     else:
```

Comment:
This part is a bit confusing to read. This else clause should come after the red highlighted part. The script is essentially the same as before save for an additional case statement if the user wants to use Morgan fingerprints.